src/main/java/net/minecraftforge/fluids/FluidType.java

```diff
@@ -0,0 +1,1128 @@
+     * @param stack the stack holding the fluid
+     * @return the {@link FluidState} being placed
+     */
+    public FluidState getStateForPlacement(BlockAndTintGetter getter, BlockPos pos, FluidStack stack)
+    {
+        return stack.getFluid().defaultFluidState();
+    }
+
+    /**
+     * Returns whether the fluid can be placed in the level.
+     *
+     * @param getter the getter which can get the level data
+     * @param pos the position of where the fluid is being placed
+     * @param state the state of the fluid being placed
+     * @return {@code true} if the fluid can be placed, {@code false} otherwise
+     */
+    public final boolean canBePlacedInLevel(BlockAndTintGetter getter, BlockPos pos, FluidState state)
+    {
+        return !this.getBlockForFluidState(getter, pos, state).isAir();
+    }
+
+    /**
+     * Returns whether the fluid can be placed in the level.
+     *
+     * @param getter the getter which can get the level data
+     * @param pos the position of where the fluid is being placed
+     * @param stack the stack holding the fluid
+     * @return {@code true} if the fluid can be placed, {@code false} otherwise
+     */
+    public final boolean canBePlacedInLevel(BlockAndTintGetter getter, BlockPos pos, FluidStack stack)
+    {
+        return this.canBePlacedInLevel(getter, pos, this.getStateForPlacement(getter, pos, stack));
+    }
+
+    /**
+     * Returns whether a fluid is lighter than air. If the fluid's density
+     * is lower than or equal {@code 0}, the fluid is considered lighter than air.
+     *
+     * <p>Tip: {@code 0} is the "canonical" density of air within Forge.
+     *
+     * <p>Note: Fluids lighter than air will have their bucket model rotated
+     * upside-down; fluid block models will have their vertices inverted.
+     *
+     * @return {@code true} if the fluid is lighter than air, {@code false} otherwise
+     */
+    public final boolean isLighterThanAir()
+    {
+        return this.getDensity() <= 0;
+    }
+
+    /**
+     * Determines if this fluid should be vaporized when placed into a level.
+     *
+     * <p>Note: Fluids that can turn lava into obsidian should vaporize within
+     * the nether to preserve the intentions of vanilla.
+     *
+     * @param level the level the fluid is being placed in
+     * @param pos the position to place the fluid at
+     * @param stack the stack holding the fluid being placed
+     * @return {@code true} if this fluid should be vaporized on placement, {@code false} otherwise
+     *
+     * @see BucketItem#emptyContents(Player, Level, BlockPos, BlockHitResult)
+     */
+    public boolean isVaporizedOnPlacement(Level level, BlockPos pos, FluidStack stack)
+    {
+        BlockState state = this.getBlockForFluidState(level, pos, this.getStateForPlacement(level, pos, stack));
+        if (state == null) return false;
+        else return level.dimensionType().ultraWarm() && state.getMaterial() == Material.WATER;
```

Comment:
I feel like checking the dimension for the default before looking up the state from the world might have slightly better performance.